README.md
# TTrack (TimeTracker)

## Disclaimer

These are my very first lines of Ruby, and this is purely meant to be a make-your-feet-wet
project, so clone this stuff only if you want to read some very ugly Ruby code

## What I ask to TimeTracker

I want to write a very simple piece of software that lets me easily keep track of time spent
on given issues. Once it works from CLI, a bonus could be to have an easy way to sync this
timesheets to some external PM software like Redmine via REST APIs

## How to use

Install TTrack

    $ gem install ttrack

Initialize sqlite3 db (defualted to $HOME/.timetrackerdb)

    $ ttrack
    $ ls ~/.timetrackerdb

Start tracking an issue (with optional notes if needed)

    $ ttrack start my_issue "my notes between quotes"

Check current tracked issue's status / time tracked

    $ ttrack status

Check current or previous issue's time tracked

    $ ttrack status my_old_issue

Stop tracking...

    $ ttrack stop

... or just start tracking a new issue, previous will be stopped automatically

    $ ttrack start my_new_issue

Get a list of tracked issues

    $ ttrack report

Get a report for a particular issue name

    $ ttrack report my_issue

Edit start and/or stop time for a given issue *number* (use UTC timestamps)

    $ ttrack report
    $ ttrack begin <issue_number> 2011-01-01 00:00
    $ ttrack end <issue_number> 2011-01-02 00:00:59


CHANGELOG
0.4.1
- add "days" to status output

0.4.0
- added "days" in start and stop output


test/ts_ttrack.rb
# -*- coding: utf-8 -*-
require 'test/unit'
require 'TTrack'
require 'FileUtils'

class TestTimeTracker < Test::Unit::TestCase

  def setup
    FileUtils.rm('testdb')  # Remove testing db before any test
    @tt = TTrack.new('testdb')
  end

  def test_init
    assert_equal @tt.class, TTrack
  end

  def test_status
    assert @tt.commands.include? :status
    assert_nil @tt.status
  end

  def test_start_stop
    assert @tt.commands.include? :start
    assert @tt.start('issuename_test', 'notes_test')
    assert @tt.commands.include? :stop
    result = @tt.stop
    assert_equal result, {:name=>'issuename_test', :id=>1, :elapsed=>0}
    assert_equal false, @tt.stop
  end

  def test_start_start_stop
    assert @tt.start('issuename_test1')
    result = @tt.start('issuename_test2')
    assert_equal true, result[0]
    assert_equal result[1], {:name=>'issuename_test1', :id=>1, :elapsed=>0}
  end

  def test_status_with_task_running
    assert @tt.start('issuename_test')
    status = @tt.status
    assert_equal 'issuename_test', status[:task]
    assert_kind_of Float, status[:elapsed]
    assert_not_equal 0, status[:elapsed]
  end

  def test_status_with_old_tasks
    assert @tt.start('issuename_test')
    assert @tt.start('issuename_test_2')
    assert @tt.start('issuename_test')
    assert @tt.start('issuename_test_2')
    status = @tt.status('issuename_test')
    assert_equal 0, status[:elapsed]
    assert_kind_of Float, status[:elapsed]
    assert_equal 'issuename_test', status[:task]
  end

  def test_status_wrong
    status = @tt.status('foo_issuename')
    assert_nil status
  end

  def test_report_void
    assert @tt.commands.include? :report
    assert_nil @tt.report
  end

  def test_report_with_name
    assert @tt.start('issuename_test')
    report = @tt.report('issuename_test')
    assert_kind_of Hash, report[0]
  end

  def test_report_with_fakename
    report = @tt.report('foo_issuename')
    assert_nil report
  end

  def test_duplicated_issuename
    assert @tt.start('issuename_test_clone')
    assert @tt.start('issuename_test_clone')
    report = @tt.report('issuename_test_clone')
    assert_kind_of Hash, report[1]
  end

  def test_cleanup_db
    assert @tt.commands.include? :init
    assert_equal [], @tt.init
    assert_nil @tt.report
  end

  def test_set_tstart
    timestamp_short = '2010-01-01 00:00'
    timestamp_full = '2010-01-01 00:00:00'
    assert @tt.start('issuename_test')
    assert @tt.set_tstart!(1, timestamp_short)
    assert_match timestamp_short, @tt.report('issuename_test')[0][:tstart]
    assert @tt.set_tstart!(1, timestamp_full)
    assert_equal timestamp_full, @tt.report('issuename_test')[0][:tstart]
  end

  def test_set_tstop
    timestamp_short = '2010-02-01 00:00'
    timestamp_full = '2010-02-01 00:00:00'
    assert @tt.start('issuename_test')
    assert @tt.set_tstop!(1, timestamp_short)
    assert_match timestamp_short, @tt.report('issuename_test')[0][:tstop]
    assert @tt.set_tstop!(1, timestamp_full)
    assert_equal timestamp_full, @tt.report('issuename_test')[0][:tstop]
  end

  def test_set_start_stop_1year
    timestamp_start = '2010-02-01 00:00:00'
    timestamp_stop = '2011-02-01 00:00:00'
    assert @tt.start('issuename_test')
    assert @tt.stop
    assert @tt.set_tstart!(1, timestamp_start)
    assert @tt.set_tstop!(1, timestamp_stop)
    assert_equal @tt.report[0][:elapsed], 31536000
  end

end


lib/ttrack.rb
# -*- coding: utf-8 -*-
#
# = TTrack class
#
# Class that handles communication between CLI and data storage
#
class TTrack

  # At each call we need to declare the datastore
  def initialize datastore
    @db = DataStore.new datastore
  end

  # Start tracking an issue, stop the current one if running
  # Return boolean to indicate if the action was succesfull or not
  def start issuename, notes=''
    unless issuename.nil? or issuename.empty?
      stopresult = stop if @db.getcurrent
      @db.startnew issuename, notes
      [true, stopresult]
    else
      false
    end
  end

  # Stop the current running task
  # Return info about the stopped running task, false if it was not tracking
  def stop
    result = @db.stoprunning
    if result and result != []
      { 
        :name => @db.getnamebyissueid(result)[0][0],
        :id => result,
        :elapsed => getissueduration(result)
      }
    else
      false
    end
  end

  # Get current running task status if no issuename given
  # Otherwise get total duration for named task (i.e. for any matching time entry)
  def status issuename=nil
    if issuename.nil?
      status = @db.getstatus
      if status
        delta = Time.now - gettstart(@db.getcurrent)
        {:task => status[1], :elapsed => delta}
      end
    else
      total = gettotalissueduration(issuename)
      unless total.nil?
        {:task => issuename, :elapsed => total}
      end
    end
  end

  # Get a report for the given issue name
  # If no issue is give, get a report for everything
  def report issuename=nil
    result = Array.new

    if issuename.nil? or issuename.empty?
      r = @db.gettimesheet
      r.each do |line|
          result << {
          :task => line[0],
          :name => line[1],
          :tstart => line[2],
          :tstop => line[3],
          :synced => line[4],
          :notes => line[5],
          :elapsed => getissueduration(line[0])
        }
      end
    else
      r = @db.getissuesbyname issuename
      r.each do |line|
          result << {
          :task => line[0],
          :tstart => line[1],
          :tstop => line[2],
          :synced => line[3],
          :notes => line[4],
          :elapsed => getissueduration(line[0])
        }
      end
    end
    result.empty? ? nil : result
  end

  # Overwrite tstart for given issue with given timestamp
  def set_tstart! issueid, timestamp
    if validateissueid(issueid)
      t = validatetimestamp(timestamp)
      @db.settstart(issueid, t)
      true
    else
      false
    end
  end

  # Override tstop for given issue
  def set_tstop! issueid, timestamp
    if validateissueid issueid
      t = validatetimestamp timestamp
      @db.settstop issueid, t
    end
  end

  # Cleanup the database
  def init
    @db.cleanup
  end

  # List of available public methods
  def commands
    [:start, :stop, :status, :init, :report, :begin, :end]
  end

  private
  def gettimeobject timestamp
    if timestamp
      Time.utc(
        timestamp[0..3],
        timestamp[5..6],
        timestamp[8..9],
        timestamp[11..12],
        timestamp[14..15],
        timestamp[17..18]
      )
    else
      Time.now
    end
  end

  def validatetimestamp timestamp
    t = Time.utc(
        timestamp[0..3],
        timestamp[5..6],
        timestamp[8..9],
        timestamp[11..12],
        timestamp[14..15],
        timestamp[17..18]
      )
    "%.4d-%.2d-%.2d %.2d:%.2d:%.2d" % [
      t.year,
      t.month,
      t.day,
      t.hour,
      t.min,
      t.sec
    ]
  end

  def validateissueid issueid
    unless @db.isissue? issueid
      puts "Issueid invalid"
      false
    else
      true
    end
  end

  def gettotalissueduration issuename
    r = @db.getissuesandtimesbyname(issuename)
    total = 0
    r.each do |timestamps|
      if timestamps[0]
        t0, t1 = gettimeobject(timestamps[0]), gettimeobject(timestamps[1])
        total = total + (t1 - t0)
      end
    end
    r[0].nil? ? nil : total
  end

  def getissueduration issueid
    timestamps = @db.gettime issueid
    t0, t1 = gettimeobject(timestamps[0]), gettimeobject(timestamps[1])
    (t1 - t0).to_int
  end

  def gettstart issueid
    timestamps = @db.gettime issueid
    unless timestamps[0].nil?
      gettimeobject timestamps[0]
    end
  end

  def gettstop issueid
    timestamps = @db.gettime issueid
    unless timestamps[1].nil?
      gettimeobject timestamps[1]
    end
  end

end  # End TTrack class

require 'ttrack/datastore'
# TODO: add support for timezone
# TODO: add version check


lib/timetracker.rb
# -*- coding: utf-8 -*-
class TimeTracker

  def initialize
    @db = DataStore.new
  end

  def start issuename, notes
    unless issuename.nil?
      if @db.getcurrent
        @db.stoprunning
      end
      @db.startnew issuename, notes
    else
      usage
    end
  end

  def stop
    @db.stoprunning
  end

  def status issuename
    unless issuename
      status = @db.getstatus
      if status
        delta = Time.now - gettstart(@db.getcurrent)
        puts "Task '%s' run time: %d seconds (%.2f hours)" % [status[1], delta, delta / 3600]
      else
        puts "Not tracking"
      end
    else
      total = gettotalissueduration(issuename)
      unless total == 0
        puts "Total seconds for task %s: %d (%.2f hours)" % 
          [
            issuename,
            total,
            total / 3600
          ]
      else
        puts "No entry for issue '%s'" % issuename
      end
    end
  end

  def report issuename
    r = @db.getissuesidsbyname issuename
    unless r[0].nil?
      puts "id | tstart | tstop | notes"
      r.each do |x|
        puts "%s | %s | %s | %s" % x
      end
    else
      usage
    end
  end

  def setstart issueid, timestamp
    if validateissueid issueid
      t = validatetimestamp timestamp
      @db.settstart issueid, t
    end
  end

  def setstop issueid, timestamp
    if validateissueid issueid
      t = validatetimestamp timestamp
      @db.settstop issueid, t
    end
  end

  def init
    @db.cleanup
  end

  def usage
    puts "Usage: timetracker [start|stop|status|init|report|begin|end] issue_name"
  end

  private
  def gettimeobject timestamp
    if timestamp
      Time.utc(
        timestamp[0..3],
        timestamp[5..6],
        timestamp[8..9],
        timestamp[11..12],
        timestamp[14..15],
        timestamp[17..18]
      )
    else
      Time.now
    end
  end

  def validatetimestamp timestamp
    t = Time.utc(
        timestamp[0..3],
        timestamp[5..6],
        timestamp[8..9],
        timestamp[11..12],
        timestamp[14..15],
        timestamp[17..18]
      )
    "%.4d-%.2d-%.2d %.2d:%.2d:%.2d" % [
      t.year,
      t.month,
      t.day,
      t.hour,
      t.min,
      t.sec
    ]
  end

  def validateissueid issueid
    unless @db.isissue? issueid
      puts "Issueid invalid"
      false
    else
      true
    end
  end

  def gettotalissueduration issuename
    r = @db.getissuesandtimesbyname(issuename)
    total = 0
    r.each do |timestamps|
      if timestamps[0]
        t0, t1 = gettimeobject(timestamps[0]), gettimeobject(timestamps[1])
        total = total + (t1 - t0)
      end
    end
    total
  end

  def getissueduration issueid
    timestamps = @db.gettime issueid
    if timestamps and timestamps[1]
      t0, t1 = gettimeobject(timestamps[0]), gettimeobject(timestamps[1])
      t1 - t0
    else
      false
    end
  end

  def gettstart issueid
    timestamps = @db.gettime issueid
    unless timestamps[0].nil?
      gettimeobject timestamps[0]
    end
  end

  def gettstop issueid
    timestamps = @db.gettime issueid
    unless timestamps[1].nil?
      gettimeobject timestamps[1]
    end
  end

end  # End TimeTracker class

require 'timetracker/datastore'
# TODO: add support for timezone


lib/ttrack/datastore.rb
# -*- coding: utf-8 -*-
require 'sqlite3'

class TTrack::DataStore

  def initialize dbname, timezone="+02:00", version='v0.3.3', verbosity=0
    @dbname = dbname
    @timezone = timezone
    @version = version
    @verbosity = verbosity

    if FileTest.zero?(@dbname) or not FileTest.file?(@dbname)
      @db = SQLite3::Database.new( @dbname )
      createschema
      init(timezone, version, verbosity)
    else
      @db = SQLite3::Database.open( @dbname )
    end
  end

  private
  def createschema
    @db.execute <<-SQL
    CREATE TABLE system (
      current,
      latest,
      timezone NOT NULL,
      version INTEGER NOT NULL,
      verbosity NOT NULL,
      redmine
    )
    SQL

    @db.execute <<-SQL
    CREATE TABLE timesheet (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      name,
      tstart DEFAULT CURRENT_TIMESTAMP,
      tstop,
      synced DEFAULT 'false',
      notes
    )
    SQL
  end

  def init timezone, version, verbosity
    @db.execute 'INSERT INTO system ( timezone, version, verbosity ) VALUES ( "%s", "%s", "%s" )' %
      [timezone, version, verbosity]
  end

  def droptables
    for table in ['system', 'timesheet']
      @db.execute 'DROP TABLE IF EXISTS %s' % table
    end
  end

  def setcurrent issueid
    oldcurrent = @db.execute 'SELECT current, latest FROM system'
    @db.execute "UPDATE system SET current = '%d', latest = '%s'" %
      [issueid, oldcurrent[0][0] != '' ? oldcurrent[0][0] : oldcurrent[0][1]]
  end

  public
  def getcurrent
    current = @db.execute "SELECT current FROM system"
    current[0][0] == "" ? false : current[0][0]
  end

  def getstatus
    current = getcurrent
    if current
      status = @db.execute "SELECT id,name FROM timesheet WHERE id=%d" % current
      status[0]
    else
      nil
    end
  end

  def startnew issuename, notes
    @db.execute "INSERT INTO timesheet ( name, notes ) VALUES ( '%s', '%s' )" %
      [issuename, notes]
    latest = @db.execute "SELECT id FROM timesheet ORDER BY id DESC LIMIT 1"
    setcurrent latest[0][0]
  end

  def stoprunning
    current = getcurrent
    if current
      @db.execute "UPDATE timesheet set tstop = datetime('now') WHERE id = %d" % current
      @db.execute "UPDATE system SET current = '', latest = '%d'" % current
      current.to_i
    else
      false
    end
  end

  def gettime issueid=false
    issueid = issueid ? issueid : getcurrent
    if issueid
      time = @db.execute "SELECT tstart,tstop FROM timesheet WHERE id=%d" % issueid
      time[0]
    else
      false
    end
  end

  def gettimesheet
    @db.execute "SELECT id,name,tstart,tstop,synced,notes FROM timesheet"
  end

  def gettimesbyissuename issuename
    @db.execute "SELECT tstart,tstop FROM timesheet WHERE name='%s'" % issuename
  end

  def getissuesandtimesbyname issuename
    @db.execute "SELECT tstart,tstop FROM timesheet WHERE name='%s'" % issuename
  end

  def getissuesbyname issuename
    @db.execute "SELECT id,tstart,tstop,synced, notes FROM timesheet WHERE name='%s'" % issuename
  end

  def getnamebyissueid issueid
    result = @db.execute "SELECT name FROM timesheet WHERE id=%d" % issueid
  end

  def settstart id, tstart
    @db.execute "UPDATE timesheet SET tstart=datetime('%s') WHERE id=%d" % [tstart, id]
  end

  def settstop id, tstop
    @db.execute "UPDATE timesheet SET tstop=datetime('%s') WHERE id=%d" % [tstop, id]
  end

  def isissue? id
    r = @db.execute "SELECT id from timesheet WHERE id=%d" % id
    r[0].nil? ? false : true
  end

  def cleanup
    droptables
    createschema
    init(@timezone, @version, @verbosity)
  end

end  # End DataStore class

main.rb
#!/Users/alex/.rvm/rubies/ruby-1.9.2-p290/bin/ruby -w
# -*- coding: utf-8 -*-
require 'sqlite3'

class TTDataStore

  def initialize dbname=".timetrackerdb", timezone="+02:00", version='v0.1', verbosity=0
    @dbname = dbname
    @timezone = timezone
    @version = version
    @verbosity = verbosity

    if FileTest.zero?(dbname) or not FileTest.file?(dbname)
      @db = SQLite3::Database.new( dbname )
      createschema
      init(timezone, version, verbosity)
    else
      @db = SQLite3::Database.open( dbname )
    end
  end

  private
  def createschema
    @db.execute <<-SQL
    CREATE TABLE system (
      current,
      latest,
      timezone NOT NULL,
      version INTEGER NOT NULL,
      verbosity NOT NULL,
      redmine
    )
    SQL

    @db.execute <<-SQL
    CREATE TABLE timesheet (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      name,
      tstart DEFAULT CURRENT_TIMESTAMP,
      tstop,
      synced DEFAULT 'false',
      notes
    )
    SQL
  end

  def init timezone, version, verbosity
    @db.execute 'INSERT INTO system ( timezone, version, verbosity ) VALUES ( "%s", "%s", "%s" )' %
      [timezone, version, verbosity]
  end

  def droptables
    for table in ['system', 'timesheet']
      @db.execute 'DROP TABLE IF EXISTS %s' % table
    end
  end

  def setcurrent issueid
    oldcurrent = @db.execute 'SELECT current, latest FROM system'
    @db.execute "UPDATE system SET current = '%d', latest = '%s'" %
      [issueid, oldcurrent[0][0] != '' ? oldcurrent[0][0] : oldcurrent[0][1]]
  end

  public
  def getcurrent
    current = @db.execute "SELECT current FROM system"
    current[0][0] == "" ? nil : current[0][0]
  end

  def getstatus
    current = getcurrent
    if current
      status = @db.execute "SELECT id,name FROM timesheet WHERE id=%d" % current
      status[0]
    else
      nil
    end
  end

  def startnew issuename, notes=''
    @db.execute "INSERT INTO timesheet ( name, notes ) VALUES ( '%s', '%s' )" %
      [issuename, notes]
    latest = @db.execute "SELECT id FROM timesheet ORDER BY id DESC LIMIT 1"
    setcurrent latest[0][0]
  end

  def stoprunning
    current = getcurrent
    if current
      @db.execute "UPDATE timesheet set tstop = datetime('now') WHERE id = %d" % current
      @db.execute "UPDATE system SET current = '', latest = '%d'" % current
    else
      p "Not tracking"
    end
  end

  def gettime issueid=false
    issueid = issueid ? issueid : getcurrent
    if issueid
      time = @db.execute "SELECT tstart,tstop FROM timesheet WHERE id=%d" % issueid
      time[0]
    else
      false
    end
  end

  def gettimesbyissuename issuename
    @db.execute "SELECT tstart,tstop FROM timesheet WHERE name='%s'" % issuename
  end

  def getissuesandtimesbyname issuename
    @db.execute "SELECT tstart,tstop FROM timesheet WHERE name='%s'" % issuename
  end

  def getissuesidsbyname issuename
    @db.execute "SELECT id,tstart,tstop,notes FROM timesheet WHERE name='%s'" % issuename
  end

  def settstart id, tstart
    @db.execute "UPDATE timesheet SET tstart=datetime('%s') WHERE id=%d" % [tstart, id]
  end

  def settstop id, tstop
    @db.execute "UPDATE timesheet SET tstop=datetime('%s') WHERE id=%d" % [tstop, id]
  end

  def isissue? id
    r = @db.execute "SELECT id from timesheet WHERE id=%d" % id
    r[0].nil? ? false : true
  end

  def cleanup
    droptables
    createschema
    init(@timezone, @version, @verbosity)
  end

end  # End DataStore class


class TimeTracker

  def initialize
    @db = TTDataStore.new
  end

  def start issuename, notes
    unless issuename.nil?
      if @db.getcurrent
        @db.stoprunning
      end
      @db.startnew issuename, notes
    else
      usage
    end
  end

  def stop
    @db.stoprunning
  end

  def status issuename
    unless issuename
      status = @db.getstatus
      if status
        delta = Time.now - gettstart(@db.getcurrent)
        p "Task '%s' run time: %d seconds (%.2f hours)" % [status[1], delta, delta / 3600]
      else
        p "Not tracking"
      end
    else
      total = gettotalissueduration(issuename)
      unless total == 0
        p "Total seconds for task %s: %d (%.2f hours)" % 
          [
            issuename,
            total,
            total / 3600
          ]
      else
        p "No entry for issue '%s'" % issuename
      end
    end
  end

  def report issuename
    r = @db.getissuesidsbyname issuename
    unless r[0].nil?
      p "id | tstart | tstop | notes"
      r.each do |x|
        p "%s | %s | %s | %s" % x
      end
    else
      usage
    end
  end

  def setstart issueid, timestamp
    if validateissueid issueid
      t = validatetimestamp timestamp
      @db.settstart issueid, t
    end
  end

  def setstop issueid, timestamp
    if validateissueid issueid
      t = validatetimestamp timestamp
      @db.settstop issueid, t
    end
  end

  def init
    @db.cleanup
  end

  def usage
    p "Usage: %s [start|stop|status|init|report|begin|end] issue_name" % $0
  end

  private
  def gettimeobject timestamp
    if timestamp
      Time.utc(
        timestamp[0..3],
        timestamp[5..6],
        timestamp[8..9],
        timestamp[11..12],
        timestamp[14..15],
        timestamp[17..18]
      )
    else
      Time.now
    end
  end

  def validatetimestamp timestamp
    t = Time.utc(
        timestamp[0..3],
        timestamp[5..6],
        timestamp[8..9],
        timestamp[11..12],
        timestamp[14..15],
        timestamp[17..18]
      )
    "%.4d-%.2d-%.2d %.2d:%.2d:%.2d" % [
      t.year,
      t.month,
      t.day,
      t.hour,
      t.min,
      t.sec
    ]
  end

  def validateissueid issueid
    unless @db.isissue? issueid
      p "Issueid invalid"
      false
    else
      true
    end
  end

  def gettotalissueduration issuename
    r = @db.getissuesandtimesbyname(issuename)
    total = 0
    r.each do |timestamps|
      if timestamps[0]
        t0, t1 = gettimeobject(timestamps[0]), gettimeobject(timestamps[1])
        total = total + (t1 - t0)
      end
    end
    total
  end

  def getissueduration issueid
    timestamps = @db.gettime issueid
    if timestamps and timestamps[1]
      t0, t1 = gettimeobject(timestamps[0]), gettimeobject(timestamps[1])
      t1 - t0
    else
      false
    end
  end

  def gettstart issueid
    timestamps = @db.gettime issueid
    unless timestamps[0].nil?
      gettimeobject timestamps[0]
    end
  end

  def gettstop issueid
    timestamps = @db.gettime issueid
    unless timestamps[1].nil?
      gettimeobject timestamps[1]
    end
  end

end  # End TimeTracker class


tt = TimeTracker.new
case ARGV[0]
  when 'start' then
    tt.start ARGV[1], ARGV[2]
  when 'stop' then
    tt.stop
  when 'status' then
    tt.status ARGV[1]
  when 'init' then
    tt.init
  when 'report' then
    tt.report ARGV[1]
  when 'beguin' then
    timestamp = ARGV[3] ? "%s %s" % ARGV[2..3] : ARGV[2]
    tt.setstart ARGV[1], timestamp
  when 'end' then
    timestamp = ARGV[3] ? "%s %s" % ARGV[2..3] : ARGV[2]
    tt.setstop ARGV[1], timestamp
  else tt.usage
end


# TODO: add support for timezone
